webrtc/videoroom/assets/src/membraneWebRTC.ts

```diff
@@ -26,9 +26,11 @@ interface TrackContext {
 }
 
 interface Callbacks {
-  onAddTrack?: (ctx: TrackContext) => void;
+  onAddTrack?: (ctx: TrackContext, display: boolean) => void;
   onRemoveTrack?: (ctx: TrackContext) => void;
   onConnectionError?: (message: string) => void;
+  onReplaceStream?: (oldStreamId: String, newStream: MediaStream) => void;
+  onDisplayStream?: (stream: MediaStream) => void;
```

Comment:
We don't know which tracks should be displayed. We only know which track should be replaced.

Reply:
Hmm ok, so what I don't like about this API is that:
- it imposes the way that the speakers are replaced, which is quite a frontend responsibility
- it is quite high-level, while we expect MembraneWebRTC to be a low-level, general-purpose utility
- it's stream-based, while the entire API is track-based

Reply:
> it imposes the way how the speakers are replaced, which is quite a frontend responsibility
it is quite high-level, while we expect MembraneWebRTC to be a low-level, general-purpose utility

These two callbacks are invoked in response to server notifications. So when server sends notification `{replace_track, track_id1, track_id2}` we invoke `onReplaceStream` as passing track here has no sense. 

I can change it but I don't exactly know what is a preffared way. There is no extra logic in membraneWebRTC.ts. It just invokes callback when notification arrives.

In the first version we had `onServerReplaceStream` which was to indicate that this callback is invoked in resposne to server demand. 

> it's stream-based, while the entire API is track-based

I am not sure why it's track-based as in fact we don't use `MediaStreamTrack` in any callback implementation.

Reply:
Also sending notification from server `replace_stream, stream_id1, stream_id2` would be a better idea but I would wait for unifying Track.id with msid etc.

Reply:
Maybe let'schange `onReplaceStream` to `onServerReplaceTrack(track1, stream1, track2, stream2)`
and `onDisplayStream` to `onServerDisplayTrack`?

Reply:
> I am not sure why it's track-based as in fact we don't use MediaStreamTrack in any callback implementation.

We have onAdd**Track** and onRemove**Track** and they accept **Track**Context, that wraps said MediaStreamTrack ;)

> we invoke onReplaceStream as passing track here has no sense

That's not that obvious, because what actually happens is that we prune some video _tracks_ and unprune others. That's why I said the API is high-level: it introduces case-specific semantics. In another use case, we may need the pruning but it will have nothing to replacing streams.

I don't insist to change that now, though I'm trying to show my point.

Reply:
> We have onAddTrack and onRemoveTrack and they accept TrackContext, that wraps said MediaStreamTrack ;)

Yeah I meant that implementation of `onAddTrack` `onRemoteTrack` etc. ignores tracks so we have track based API but we don't use these tracks anywhere.
 
> I don't insist to change that now, though I'm trying to show my point.
I see. I mean I know this might not look perfect but as long as changing this doesn't require any changes on the server side (i.e. we can say that server side API is ok)  I would leave it.

Reply:
FYI the track-based API comes from WebRTC API, which was stream-based, but they made it track-based. Let's leave it then, but I'd rethink it when we get back to limiting audio.

Reply:
The WebRTC may be track based but what is passed to video element is still a media stream and there is no way to directly add tracks to it without changing passed stream.